noxfile.py

```diff
@@ -36,28 +75,106 @@ def run_readable(session, mode="fmt"):
         "ghcr.io/bobheadxi/readable:v0.5.0@sha256:423c133e7e9ca0ac20b0ab298bd5dbfa3df09b515b34cbfbbe8944310cc8d9c9",
         mode,
         "![.]**/*.md",
+        external=True,
     )
 
 
+def run_shellcheck(session, mode="check"):
+    shellcheck_cmd = [
+        "docker",
+        "run",
+        "--platform",
+        "linux/amd64",  # while this image is multi-arch, we cannot use digest with multi-arch images
+        "--rm",
+        "-v",
+        f"{ROOT.absolute()}:/mnt",
+        "-w",
+        "/mnt",
+        "-q",
+        "koalaman/shellcheck:0.9.0@sha256:a527e2077f11f28c1c1ad1dc784b5bc966baeb3e34ef304a0ffa72699b01ad9c",
+    ]
+
+    files = list_files(suffix=".sh")
+    if not files:
+        session.log("No shell files found")
+        return
+    shellcheck_cmd.extend(files)
+
+    if mode == "fmt":
+        with tempfile.NamedTemporaryFile(mode="w+") as diff_file:
+            session.run(
+                *shellcheck_cmd,
+                "--format=diff",
+                external=True,
+                stdout=diff_file,
+                success_codes=[0, 1],
+            )
+            diff_file.seek(0)
+            diff = diff_file.read()
+            if len(diff.splitlines()) > 1:  # ignore single-line message
+                session.log("Applying shellcheck patch:\n%s", diff)
+                subprocess.run(
+                    ["patch", "-p1"],
+                    input=diff,
+                    text=True,
+                    check=True,
+                )
+
+    session.run(*shellcheck_cmd, external=True)
+
+
 @nox.session(name="format", python=PYTHON_DEFAULT_VERSION)
 def format_(session):
-    session.run_always("pdm", "install", "--dev")
-    session.run("black", ".")
+    """Lint the code and apply fixes in-place whenever possible."""
+    install(session, "lint")
     session.run("ruff", "check", "--fix", ".")
+    run_shellcheck(session, mode="fmt")
```

Comment:
suggestion: I found really anoying that in single session we have multiple checks that has to run in single virtualenv - in https://github.com/backend-developers-ltd/compute-horde-facilitator-sdk/blob/master/noxfile.py I've used tags to control which sessions to select in nox run and then you can run single session or all sessions under choosen tag

Reply:
can you explain why did you find it annoying?

I personally disliked the fact, that I had to manually analyze each project just to know which commands I need to run before its safe to create a PR that won't error out on CI stage.

I just want a single command that I will run and I will be decently sure that my PR will pass CI.

I don't want to guess whenever mypy is needed or not for example.

Reply:
don't remember exactly what was the nail... :) but I wanted to have clean separate virtualenvs for different linters and be able to select one linter and run it - so sessions be like commands and tags be like sets of commands

it might be, that one thing was breaking and it was annoying to have to wait for all other linters to find out that still something is not working

tags works the same as sessions right now - just you can split linters per session, have seprate small virtualenvs and run single linter if you like

Reply:
and you can define default tags
> nox.options.tags = ["check"]

Reply:
most linters will have to run in the same env tho, since any linter that requires to analyze whenever your imports make sense or type hints are correct requires all 3rd party stuff installed

also installing&activating N venvs rather 1 will be more resource intensive, ot less

Reply:
it was that running some docker thingy was taking some time and I needed to run the linter after this and was unable - but still - you can setup single venv and reuse between sessions :) but still - it is a suggestion that I came up with when trying to work with our projects and got frustrated much and that is why I don't use nox just "raw" ruff... I will need still doing it - not using nox :P

Reply:
To facilitate tags use in future I will add the "lint" and "format" tags and recommend calling `nox -t lint` in the README, so we are at least half way there.

Reply:
if it runs 1s that we have nothing to discuss - but it was taking like 60s on gtihub actions and also it was not this fast locally - but maybe it was misconfigured